File: src/main/scala/com/mblinn/oo/iterator/HigherOrderFunctions.scala
package com.mblinn.oo.iterator

object HigherOrderFunctions {

  val isVowel = Set('a', 'e', 'i', 'o', 'u')

  def vowelsInWord(word: String) = word.filter(isVowel).toSet

  def sumSequence(sequence: Seq[Int]) =
    if (sequence.isEmpty) 0 else sequence.reduce((acc, curr) => acc + curr)

  def prependHello(names: Seq[String]) =
    names.map((name) => "Hello, " + name)
}

File: src/main/java/com/mblin/oo/iterator/HigherOrderFunctions.java
package com.mblin.oo.iterator;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class HigherOrderFunctions {
    private static final Set<Character> VOWELS = Set.of('a', 'e', 'i', 'o', 'u');

    public static Integer sumSequence(List<Integer> sequence) {
        return sequence.stream().reduce(Integer::sum).orElse(0);
    }

    public static List<String> prependHello(List<String> names) {
        return names.stream().map(n -> "Hello, " + n).collect(Collectors.toList());
    }

    public static Set<Character> vowelsInWord(String word) {
        var vowelsInWord = new HashSet<Character>();
        for (Character character : word.toLowerCase().toCharArray()) {
            if (VOWELS.contains(character)) {
                vowelsInWord.add(character);
            }
        }
        return vowelsInWord;
    }
}